packages/core/src/connectivity/scp-cf/jwt.ts

```diff
@@ -178,30 +207,41 @@ const defaultVerifyJwtOptions: VerifyJwtOptions = {
 
 function getVerificationKey(
   xsuaaCredentials: XsuaaServiceCredentials,
-  jku: string
+  header: JwtHeader
 ): Promise<TokenKey> {
-  return fetchVerificationKeys(xsuaaCredentials, jku).then(verificationKeys => {
-    if (!verificationKeys.length) {
-      throw Error(
-        'No verification keys have been returned by the XSUAA service.'
+  return fetchVerificationKeys(xsuaaCredentials, header.jku).then(
+    verificationKeys => {
+      if (!verificationKeys.length) {
+        throw Error(
+          'No verification keys have been returned by the XSUAA service.'
+        );
+      }
+      const verificationKey = verificationKeys.find(
+        key => key.keyId === header.kid
       );
+      if (!verificationKey) {
+        throw new Error(
+          'Could not find verification key for the given key ID.'
+        );
+      }
+      return verificationKey;
     }
-    return verificationKeys[0];
-  });
+  );
 }
 
 // 15 minutes is the default value used by the xssec lib
-export const verificationKeyCache = new Cache({ minutes: 15 });
+export const verificationKeyCache = new Cache<TokenKey>({ minutes: 15 });
 
-function cacheVerificationKey(
-  verificationKeyURL: string,
-  key: TokenKey,
-  options: VerifyJwtOptions
-): TokenKey {
-  if (options.cacheVerificationKeys) {
-    verificationKeyCache.set(verificationKeyURL, key);
+function buildCacheKey(
+  jku: string | undefined,
+  kid: string | undefined
+): string {
+  if (!jku || !kid) {
```

Comment:
In the  validateJwtHeaderForVerification it is already checked if the `jku` and `kid` are there. Is this second check neede?

Reply:
Yes, because the types we pass here are optional properties. This error message will likely never be reached, but I thought just to be safe it is better to check it here.